File: app/src/main/java/com/proxerme/app/util/helper/MaterialDrawerHelper.java
package com.proxerme.app.util.helper;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.Toolbar;
import android.view.View;

import com.mikepenz.community_material_typeface_library.CommunityMaterial;
import com.mikepenz.materialdrawer.AccountHeader;
import com.mikepenz.materialdrawer.AccountHeaderBuilder;
import com.mikepenz.materialdrawer.Drawer;
import com.mikepenz.materialdrawer.DrawerBuilder;
import com.mikepenz.materialdrawer.holder.BadgeStyle;
import com.mikepenz.materialdrawer.holder.StringHolder;
import com.mikepenz.materialdrawer.model.PrimaryDrawerItem;
import com.mikepenz.materialdrawer.model.ProfileDrawerItem;
import com.mikepenz.materialdrawer.model.ProfileSettingDrawerItem;
import com.mikepenz.materialdrawer.model.interfaces.IDrawerItem;
import com.mikepenz.materialdrawer.model.interfaces.IProfile;
import com.proxerme.app.R;
import com.proxerme.app.application.MainApplication;
import com.proxerme.app.manager.UserManager;
import com.proxerme.library.connection.UrlHolder;
import com.proxerme.library.entity.LoginUser;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.ArrayList;

/**
 * A helper class for managing the MaterialDrawer.
 *
 * @author Ruben Gees
 */
public class MaterialDrawerHelper {

    public static final int DRAWER_ID_NONE = -1;

    //Normal items
    public static final int DRAWER_ID_NEWS = 0;
    public static final int DRAWER_ID_MESSAGES = 1;

    //Sticky items
    public static final int DRAWER_ID_DONATE = 10;
    public static final int DRAWER_ID_SETTINGS = 11;

    //Header items
    public static final int HEADER_ID_GUEST = 100;
    public static final int HEADER_ID_USER = 101;
    public static final int HEADER_ID_LOGIN = 111;
    public static final int HEADER_ID_CHANGE = 112;
    public static final int HEADER_ID_LOGOUT = 113;

    private static final String STATE_CURRENT_DRAWER_ITEM_ID = "drawer_current_drawer_item_id";

    private AccountHeader header;
    private Drawer drawer;

    private UserManager userManager;

    private int currentDrawerItemId = -1;

    private MaterialDrawerCallback callback;

    private Drawer.OnDrawerItemClickListener onDrawerItemClickListener =
            new Drawer.OnDrawerItemClickListener() {
                @Override
                public boolean onItemClick(View view, int position, IDrawerItem iDrawerItem) {
                    int id = (int) iDrawerItem.getIdentifier();

                    if (id != currentDrawerItemId) {
                        if (iDrawerItem.isSelectable()) {
                            currentDrawerItemId = id;
                        }

                        if (callback != null) {
                            //noinspection WrongConstant
                            return callback.onItemClick(id);
                        }
                    }

                    return true;
                }
            };

    private AccountHeader.OnAccountHeaderListener onAccountHeaderClickListener =
            new AccountHeader.OnAccountHeaderListener() {
                @Override
                public boolean onProfileChanged(View view, IProfile profile, boolean current) {
                    //noinspection WrongConstant
                    return callback != null &&
                            callback.onAccountClick((int) profile.getIdentifier());
                }
            };

    public MaterialDrawerHelper(@NonNull Activity context,
                                @Nullable MaterialDrawerCallback callback) {
        this.callback = callback;

        this.userManager = ((MainApplication) context.getApplication()).getUserManager();
    }

    public void build(@NonNull Activity context, @NonNull Toolbar toolbar,
                      @Nullable Bundle savedInstanceState) {
        if (savedInstanceState != null) {
            currentDrawerItemId = savedInstanceState.getInt(STATE_CURRENT_DRAWER_ITEM_ID);
        }

        initHeader(context, savedInstanceState);
        initDrawer(context, toolbar, savedInstanceState);
    }

    private void initHeader(@NonNull Activity context, Bundle savedInstanceState) {
        header = new AccountHeaderBuilder()
                .withHeaderBackground(R.color.colorAccent)
                .withActivity(context)
                .withSavedInstance(savedInstanceState)
                .withProfiles(generateProfiles(context))
                .withOnAccountHeaderListener(onAccountHeaderClickListener)
                .build();
    }

    private ArrayList<IProfile> generateProfiles(@NonNull Activity context) {
        LoginUser user = userManager.getUser();
        ArrayList<IProfile> result = new ArrayList<>();

        if (user == null) {
            result.add(new ProfileDrawerItem()
                    .withName(context.getString(R.string.drawer_profile_guest))
                    .withIcon(R.mipmap.ic_launcher)
                    .withIdentifier(HEADER_ID_GUEST));
            result.add(new ProfileSettingDrawerItem()
                    .withName(context.getString(R.string.drawer_header_login))
                    .withIconTinted(true)
                    .withIcon(CommunityMaterial.Icon.cmd_account_key)
                    .withIdentifier(HEADER_ID_LOGIN));
        } else {
            ProfileDrawerItem profile = new ProfileDrawerItem().withName(user.getUsername())
                    .withIdentifier(HEADER_ID_USER);

            profile.withIcon(UrlHolder.getUserImageUrl(user.getImageId()));

            result.add(profile);
            result.add(new ProfileSettingDrawerItem()
                    .withName(context.getString(R.string.drawer_header_change))
                    .withIcon(CommunityMaterial.Icon.cmd_account_switch)
                    .withIconTinted(true)
                    .withIdentifier(HEADER_ID_CHANGE));
            result.add(new ProfileSettingDrawerItem()
                    .withName(context.getString(R.string.drawer_header_logout))
                    .withIcon(CommunityMaterial.Icon.cmd_account_remove)
                    .withIconTinted(true)
                    .withIdentifier(HEADER_ID_LOGOUT));
        }

        return result;
    }

    private void initDrawer(@NonNull Activity context, @NonNull Toolbar toolbar,
                            @Nullable Bundle savedInstanceState) {
        drawer = new DrawerBuilder(context)
                .withAccountHeader(header)
                .withDrawerItems(generateDrawerItems())
                .withStickyDrawerItems(generateStickyDrawerItems())
                .withOnDrawerItemClickListener(onDrawerItemClickListener)
                .withShowDrawerOnFirstLaunch(true).withToolbar(toolbar)
                .withActionBarDrawerToggleAnimated(true)
                .withTranslucentStatusBar(true)
                .withSavedInstance(savedInstanceState).build();
    }

    @NonNull
    private ArrayList<IDrawerItem> generateDrawerItems() {
        ArrayList<IDrawerItem> result = new ArrayList<>(2);

        result.add(new PrimaryDrawerItem().withName(R.string.drawer_item_news)
                .withIcon(CommunityMaterial.Icon.cmd_newspaper)
                .withSelectedTextColorRes(R.color.colorPrimary)
                .withSelectedIconColorRes(R.color.colorPrimary)
                .withBadgeStyle(new BadgeStyle()
                        .withColorRes(R.color.colorPrimary)
                        .withTextColorRes(android.R.color.white))
                .withIdentifier(DRAWER_ID_NEWS));

        result.add(new PrimaryDrawerItem().withName(R.string.drawer_item_messages)
                .withIcon(CommunityMaterial.Icon.cmd_message_text)
                .withSelectedTextColorRes(R.color.colorPrimary)
                .withSelectedIconColorRes(R.color.colorPrimary)
                .withBadgeStyle(new BadgeStyle()
                        .withColorRes(R.color.colorPrimary)
                        .withTextColorRes(android.R.color.white))
                .withIdentifier(DRAWER_ID_MESSAGES));

        return result;
    }

    @NonNull
    private ArrayList<IDrawerItem> generateStickyDrawerItems() {
        ArrayList<IDrawerItem> result = new ArrayList<>(2);

        result.add(new PrimaryDrawerItem().withName(R.string.drawer_item_donate)
                .withIcon(CommunityMaterial.Icon.cmd_gift)
                .withSelectedTextColorRes(R.color.colorPrimary)
                .withSelectedIconColorRes(R.color.colorPrimary)
                .withSelectable(false)
                .withIdentifier(DRAWER_ID_DONATE));

        result.add(new PrimaryDrawerItem().withName(R.string.drawer_item_settings)
                .withIcon(CommunityMaterial.Icon.cmd_settings)
                .withSelectedTextColorRes(R.color.colorPrimary)
                .withSelectedIconColorRes(R.color.colorPrimary)
                .withIdentifier(DRAWER_ID_SETTINGS));

        return result;
    }

    public void saveInstanceState(Bundle outState) {
        outState.putInt(STATE_CURRENT_DRAWER_ITEM_ID, currentDrawerItemId);
        header.saveInstanceState(outState);
        drawer.saveInstanceState(outState);
    }

    public boolean isDrawerOpen() {
        return drawer.isDrawerOpen();
    }

    public void select(@DrawerItemId int id) {
        drawer.setSelection(id);
    }

    public boolean handleBackPressed() {
        if (isDrawerOpen()) {
            drawer.closeDrawer();

            return true;
        } else if (currentDrawerItemId != DRAWER_ID_NEWS) {
            select(DRAWER_ID_NEWS);

            return true;
        } else {
            return false;
        }
    }

    public void setBadge(@DrawerItemId int drawerItemId, @Nullable String text) {
        if (text == null) {
            drawer.updateBadge(drawerItemId, null);
        } else {
            drawer.updateBadge(drawerItemId, new StringHolder(text));
        }
    }

    public void refreshHeader(@NonNull Activity context) {
        header.setProfiles(generateProfiles(context));
    }

    @Retention(RetentionPolicy.SOURCE)
    @IntDef({DRAWER_ID_NONE, DRAWER_ID_NEWS, DRAWER_ID_MESSAGES, DRAWER_ID_DONATE,
            DRAWER_ID_SETTINGS})
    public @interface DrawerItemId {
    }

    @Retention(RetentionPolicy.SOURCE)
    @IntDef({HEADER_ID_GUEST, HEADER_ID_USER, HEADER_ID_LOGIN, HEADER_ID_LOGOUT,
            HEADER_ID_CHANGE})
    public @interface HeaderItemId {
    }

    public static abstract class MaterialDrawerCallback {
        public boolean onItemClick(@DrawerItemId int identifier) {
            return false;
        }

        public boolean onAccountClick(@HeaderItemId int identifier) {
            return false;
        }
    }
}

File: app/src/main/kotlin/com/proxerme/app/helper/MaterialDrawerHelper.kt
package com.proxerme.app.helper

import android.app.Activity
import android.os.Bundle
import android.support.v7.widget.Toolbar
import android.view.View
import com.mikepenz.community_material_typeface_library.CommunityMaterial
import com.mikepenz.crossfader.Crossfader
import com.mikepenz.crossfader.util.UIUtils
import com.mikepenz.crossfader.view.GmailStyleCrossFadeSlidingPaneLayout
import com.mikepenz.materialdrawer.*
import com.mikepenz.materialdrawer.holder.BadgeStyle
import com.mikepenz.materialdrawer.model.PrimaryDrawerItem
import com.mikepenz.materialdrawer.model.ProfileDrawerItem
import com.mikepenz.materialdrawer.model.ProfileSettingDrawerItem
import com.mikepenz.materialdrawer.model.interfaces.IDrawerItem
import com.mikepenz.materialdrawer.model.interfaces.IProfile
import com.proxerme.app.R
import com.proxerme.app.util.CrossfadeWrapper
import com.proxerme.app.util.DeviceUtils
import com.proxerme.library.info.ProxerUrlHolder

/**
 * TODO: Describe class
 *
 * @author Ruben Gees
 */
class MaterialDrawerHelper(context: Activity, toolbar: Toolbar,
                           savedInstanceState: Bundle?,
                           private val itemClickCallback: (id: DrawerItem) -> Boolean = { false },
                           private val accountClickCallback: (id: AccountItem) -> Boolean = { false }) {

    companion object {
        private const val CROSSFADER_FIRST_DRAWER_DP: Float = 300f
        private const val CROSSFADER_SECOND_DRAWER_DP: Float = 72f
        private const val STATE_CURRENT_DRAWER_ITEM_ID = "material_drawer_helper_current_id"
    }

    private val header: AccountHeader
    private val drawer: Drawer

    private var miniDrawer: MiniDrawer?
    private var crossfader: Crossfader<*>?

    private var currentItem: DrawerItem? = null

    init {
        header = buildAccountHeader(context, savedInstanceState)
        drawer = buildDrawer(context, toolbar, header, savedInstanceState)

        val tabletDrawer = buildTabletDrawer(context, drawer, savedInstanceState)
        miniDrawer = tabletDrawer.first
        crossfader = tabletDrawer.second

        currentItem = DrawerItem.fromOrNull(savedInstanceState?.getLong(STATE_CURRENT_DRAWER_ITEM_ID))
    }

    fun onBackPressed(): Boolean {
        if (crossfader?.isCrossFaded() ?: false) {
            crossfader?.crossFade()

            return true
        } else if (isDrawerOpen()) {
            drawer.closeDrawer()

            return true
        } else {
            val context = when {
                drawer.drawerLayout != null -> drawer.drawerLayout.context
                crossfader != null -> crossfader?.getCrossFadeSlidingPaneLayout()?.context
                else -> null
            } ?: return false

            val startPage = PreferenceHelper.getStartPage(context)

            if (currentItem != startPage) {
                select(startPage)

                return true
            } else {
                return false
            }
        }
    }

    fun saveInstanceState(outState: Bundle) {
        outState.putLong(STATE_CURRENT_DRAWER_ITEM_ID, currentItem?.id ?: DrawerItem.NEWS.id)

        header.saveInstanceState(outState)
        drawer.saveInstanceState(outState)
        crossfader?.saveInstanceState(outState)
    }

    fun isDrawerOpen(): Boolean {
        return drawer.isDrawerOpen
    }

    fun select(item: DrawerItem) {
        drawer.setSelection(item.id)
        miniDrawer?.setSelection(item.id)
    }

    fun refreshHeader(context: Activity) {
        header.profiles = generateAccountItems(context)
        drawer.recyclerView.adapter.notifyDataSetChanged()
        miniDrawer?.createItems()
    }

    private fun buildAccountHeader(context: Activity, savedInstanceState: Bundle?): AccountHeader {
        return AccountHeaderBuilder()
                .withActivity(context)
                .withCompactStyle(true)
                .withHeaderBackground(R.color.colorPrimary)
                .withOnAccountHeaderListener { view, profile, current ->
                    onAccountItemClick(view, profile, current)
                }
                .withSavedInstance(savedInstanceState)
                .withProfiles(generateAccountItems(context))
                .build()
    }

    private fun generateAccountItems(context: Activity): List<IProfile<*>> {
        val user = StorageHelper.user

        when (user) {
            null -> return arrayListOf(
                    ProfileDrawerItem()
                            .withName(context.getString(R.string.drawer_account_guest))
                            .withIcon(R.mipmap.ic_launcher)
                            .withSelectedTextColorRes(R.color.colorAccent)
                            .withIdentifier(AccountItem.GUEST.id),
                    ProfileSettingDrawerItem()
                            .withName(context.getString(R.string.drawer_account_login))
                            .withIcon(CommunityMaterial.Icon.cmd_account_key)
                            .withIconTinted(true)
                            .withIdentifier(AccountItem.LOGIN.id))
            else -> return arrayListOf(
                    ProfileDrawerItem()
                            .withName(user.username)
                            .withEmail(context.getString(R.string.login_state_indicator_logged_in))
                            .withIcon(ProxerUrlHolder.getUserImageUrl(user.imageId).toString())
                            .withSelectedTextColorRes(R.color.colorAccent)
                            .withIdentifier(AccountItem.USER.id),
                    ProfileSettingDrawerItem()
                            .withName(context.getString(R.string.drawer_account_ucp))
                            .withIcon(CommunityMaterial.Icon.cmd_account_key)
                            .withIconTinted(true)
                            .withIdentifier(AccountItem.UCP.id),
                    ProfileSettingDrawerItem()
                            .withName(context.getString(R.string.drawer_account_logout))
                            .withIcon(CommunityMaterial.Icon.cmd_account_remove)
                            .withIconTinted(true)
                            .withIdentifier(AccountItem.LOGOUT.id)
            )
        }
    }

    private fun buildDrawer(context: Activity, toolbar: Toolbar, accountHeader: AccountHeader,
                            savedInstanceState: Bundle?): Drawer {
        val builder = DrawerBuilder(context)
                .withToolbar(toolbar)
                .withAccountHeader(accountHeader)
                .withDrawerItems(generateDrawerItems())
                .withStickyDrawerItems(generateStickyDrawerItems())
                .withOnDrawerItemClickListener { view, id, item ->
                    onDrawerItemClick(view, id, item)
                }
                .withShowDrawerOnFirstLaunch(true)
                .withTranslucentStatusBar(true)
                .withGenerateMiniDrawer(DeviceUtils.isTablet(context))
                .withSavedInstance(savedInstanceState)

        return if (DeviceUtils.isTablet(context)) builder.buildView() else builder.build()
    }

    private fun buildTabletDrawer(context: Activity, drawer: Drawer, savedInstanceState: Bundle?):
            Pair<MiniDrawer?, Crossfader<*>?> {
        if (DeviceUtils.isTablet(context)) {
            val miniDrawer = drawer.miniDrawer.withIncludeSecondaryDrawerItems(true)
            val crossfader = buildCrossfader(context, drawer, miniDrawer, savedInstanceState)

            miniDrawer.withCrossFader(CrossfadeWrapper(crossfader))
            crossfader.getCrossFadeSlidingPaneLayout()?.setShadowResourceLeft(R.drawable.material_drawer_shadow_left)

            return miniDrawer to crossfader
        }

        return null to null
    }

    private fun buildCrossfader(context: Activity, drawer: Drawer, miniDrawer: MiniDrawer,
                                savedInstanceState: Bundle?): Crossfader<*> {
        return Crossfader<GmailStyleCrossFadeSlidingPaneLayout>()
                .withContent(context.findViewById(R.id.drawerRoot))
                .withFirst(drawer.slider,
                        Math.round(UIUtils.convertDpToPixel(CROSSFADER_FIRST_DRAWER_DP, context)))
                .withSecond(miniDrawer.build(context),
                        Math.round(UIUtils.convertDpToPixel(CROSSFADER_SECOND_DRAWER_DP, context)))
                .withSavedInstance(savedInstanceState)
                .build()
    }

    private fun generateDrawerItems(): List<IDrawerItem<*, *>> {
        return arrayListOf(
                PrimaryDrawerItem()
                        .withName(R.string.drawer_item_news)
                        .withIcon(CommunityMaterial.Icon.cmd_newspaper)
                        .withSelectedTextColorRes(R.color.colorAccent)
                        .withSelectedIconColorRes(R.color.colorAccent)
                        .withBadgeStyle(BadgeStyle()
                                .withColorRes(R.color.colorAccent)
                                .withTextColorRes(android.R.color.white))
                        .withIdentifier(DrawerItem.NEWS.id),
                PrimaryDrawerItem()
                        .withName(R.string.drawer_item_chat)
                        .withIcon(CommunityMaterial.Icon.cmd_message_text)
                        .withSelectedTextColorRes(R.color.colorAccent)
                        .withSelectedIconColorRes(R.color.colorAccent)
                        .withBadgeStyle(BadgeStyle()
                                .withColorRes(R.color.colorAccent)
                                .withTextColorRes(android.R.color.white))
                        .withIdentifier(DrawerItem.CHAT.id),
                PrimaryDrawerItem()
                        .withName(R.string.drawer_item_reminder)
                        .withIcon(CommunityMaterial.Icon.cmd_bookmark)
                        .withSelectedTextColorRes(R.color.colorAccent)
                        .withSelectedIconColorRes(R.color.colorAccent)
                        .withBadgeStyle(BadgeStyle()
                                .withColorRes(R.color.colorAccent)
                                .withTextColorRes(android.R.color.white))
                        .withIdentifier(DrawerItem.REMINDER.id),
                PrimaryDrawerItem()
                        .withName(R.string.drawer_item_anime)
                        .withIcon(CommunityMaterial.Icon.cmd_television)
                        .withSelectedTextColorRes(R.color.colorAccent)
                        .withSelectedIconColorRes(R.color.colorAccent)
                        .withBadgeStyle(BadgeStyle()
                                .withColorRes(R.color.colorAccent)
                                .withTextColorRes(android.R.color.white))
                        .withIdentifier(DrawerItem.ANIME.id),
                PrimaryDrawerItem()
                        .withName(R.string.drawer_item_manga)
                        .withIcon(CommunityMaterial.Icon.cmd_book_open_variant)
                        .withSelectedTextColorRes(R.color.colorAccent)
                        .withSelectedIconColorRes(R.color.colorAccent)
                        .withBadgeStyle(BadgeStyle()
                                .withColorRes(R.color.colorAccent)
                                .withTextColorRes(android.R.color.white))
                        .withIdentifier(DrawerItem.MANGA.id)
        )
    }

    private fun generateStickyDrawerItems(): ArrayList<IDrawerItem<*, *>> {
        return arrayListOf(
                PrimaryDrawerItem()
                        .withName(R.string.drawer_item_info)
                        .withIcon(CommunityMaterial.Icon.cmd_information_outline)
                        .withSelectedTextColorRes(R.color.colorAccent)
                        .withSelectedIconColorRes(R.color.colorAccent)
                        .withIdentifier(DrawerItem.INFO.id),
                PrimaryDrawerItem()
                        .withName(R.string.drawer_item_donate)
                        .withIcon(CommunityMaterial.Icon.cmd_gift)
                        .withSelectedTextColorRes(R.color.colorAccent)
                        .withSelectedIconColorRes(R.color.colorAccent)
                        .withSelectable(false)
                        .withIdentifier(DrawerItem.DONATE.id),
                PrimaryDrawerItem()
                        .withName(R.string.drawer_item_settings)
                        .withIcon(CommunityMaterial.Icon.cmd_settings)
                        .withSelectedTextColorRes(R.color.colorAccent)
                        .withSelectedIconColorRes(R.color.colorAccent)
                        .withIdentifier(DrawerItem.SETTINGS.id))
    }

    private fun getStickyItemIds(): Array<DrawerItem> {
        return generateStickyDrawerItems()
                .mapNotNull { DrawerItem.fromOrNull(it.identifier) }
                .toTypedArray()
    }

    @Suppress("UNUSED_PARAMETER")
    private fun onDrawerItemClick(view: View?, id: Int, item: IDrawerItem<*, *>): Boolean {
        if (item.identifier != currentItem?.id) {
            val newItem = DrawerItem.fromOrDefault(item.identifier)

            if (item.isSelectable) {
                currentItem = newItem

                if (miniDrawer != null && newItem in getStickyItemIds()) {
                    miniDrawer?.adapter?.deselect()
                }
            }

            return itemClickCallback.invoke(newItem)
        }

        return true
    }

    @Suppress("UNUSED_PARAMETER")
    private fun onAccountItemClick(view: View?, profile: IProfile<*>, current: Boolean): Boolean {
        return accountClickCallback.invoke(AccountItem.fromOrDefault(profile.identifier))
    }

    enum class DrawerItem(val id: Long) {
        NEWS(0L),
        CHAT(1L),
        REMINDER(2L),
        ANIME(3L),
        MANGA(4L),
        INFO(10L),
        DONATE(11L),
        SETTINGS(12L);

        companion object {
            fun fromOrNull(id: Long?) = values().firstOrNull { it.id == id }
            fun fromOrDefault(id: Long?) = fromOrNull(id) ?: NEWS
        }
    }

    enum class AccountItem(val id: Long) {
        GUEST(100L),
        LOGIN(101L),
        USER(102L),
        LOGOUT(103L),
        UCP(104L);

        companion object {
            fun fromOrNull(id: Long?) = values().firstOrNull { it.id == id }
            fun fromOrDefault(id: Long?) = fromOrNull(id) ?: USER
        }
    }
}
